### PR TITLE
Created basic compute interface

### PR DESCRIPTION
**Pull request by oierlauzi**
Created the compute interface. This interface enables basic interactions with compute devices, such as GPUs. This includes:

- Querying available device backends (CUDA, host...)
- Querying available devices from each backend (or collectively)
- Query device properties.
- Create command queues for a given device.
- Allocate memory local to a device.
- Transfer data from and to devices asynchronously in queues. In addition, for unified memory architectures (tipically integrated GPUs), data transfers can occur in a zero-copy manner.

### CODE REVIEW

**Comment by oierlauzi** [comment]
I will try to create a couple of implementations of these interfaces (host and CUDA) before marking ready for review

File: include/xmipp4/core/compute/device_index.hpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+     * @return const std::string& The name.
+     */
+    const std::string& get_backend_name() const noexcept;
+
+    /**
+     * @brief Set the ID of the device within the backend.
+     * 
+     * @param device_id The device ID.
+     */
+    void set_device_id(std::size_t device_id) noexcept;
```
Comment (MartinSalinas98) [inline include/xmipp4/core/compute/device_index.hpp:88]: Remove setter: not needed

File: include/xmipp4/core/compute/device_index.hpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    device_index& operator=(device_index &&other) = default; 
+
+    /**
+     * @brief Set the name of the device backend
+     * 
+     * @tparam Str String type,
+     * @param backend_name Name of the backend.
+     */
+    template <typename Str>
+    void set_backend_name(Str &&backend_name);
```
Comment (MartinSalinas98) [inline include/xmipp4/core/compute/device_index.hpp:74]: Remove setter: not needed

**Review by Ratolon — DISMISSED** [review]
Not much to say, looks like a complete enough version for a first compute backend onto which build. I'd say approve but @MartinSalinas98 please di algo.

File: include/xmipp4/core/compute/numerical_type.inl [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    case numerical_type::int64: return 8;
+    case numerical_type::uint64: return 8;
+    case numerical_type::float16: return 2;
+    case numerical_type::brain_float16: return 2;
+    case numerical_type::float32: return 4;
+    case numerical_type::float64: return 8;
+    case numerical_type::complex_float16: return 4;
+    case numerical_type::complex_float32: return 8;
+    case numerical_type::complex_float64: return 16;
+    default: return 0;
```
Comment (MartinSalinas98) [inline include/xmipp4/core/compute/numerical_type.inl:51]: Change magic numbers by sizeof when possible.

File: include/xmipp4/core/system/host.hpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+std::string get_hostname();
+
+/**
+ * @brief Get the total memory available in the system.
+ * 
+ * @return std::size_t Total memory in bytes.
+ */
+std::size_t get_total_system_memory();
+
```
Comment (MartinSalinas98) [inline include/xmipp4/core/system/host.hpp:51]: Add `get_available_system_memory` method. Makes sense to have that information

File: include/xmipp4/core/compute/device_properties.hpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    device_type m_type;
+    std::string m_name;
+    std::string m_physical_location;
+    std::size_t m_total_memory_bytes;
+
+};
+
+} // namespace compute
+} // namespace xmipp4
+
```
Comment (MartinSalinas98) [inline include/xmipp4/core/compute/device_properties.hpp:123]: Add method `get_available_memory`

File: src/system/host.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        MEMORYSTATUSEX status;
+        status.dwLength = sizeof(status);
+        GlobalMemoryStatusEx(&status);
+        return status.ullTotalPhys;
+    #else
+        #pragma message ("Cannot get total memory for this platform")
+        return "";
+    #endif
+}
+
```
Comment (MartinSalinas98) [inline src/system/host.cpp:78]: `get_available_system_memory`

File: tests/unitary/src/compute/test_device_index.cpp [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    REQUIRE( index.get_device_id() == 4 );
+}
+
+TEST_CASE( "set device backend in device index", "[device_index]" )
+{
+    device_index index("example-backend", 6);
+    index.set_backend_name("new-backend");
+    REQUIRE( index.get_backend_name() == "new-backend" );
+    REQUIRE( index.get_device_id() == 6 );
+}
```
Comment (MartinSalinas98) [inline tests/unitary/src/compute/test_device_index.cpp:51]: Remove setter

File: tests/unitary/src/compute/test_device_index.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+TEST_CASE( "set device id in device index", "[device_index]" )
+{
+    device_index index("example-backend", 6);
+    index.set_device_id(9);
+    REQUIRE( index.get_backend_name() == "example-backend" );
+    REQUIRE( index.get_device_id() == 9 );
+}
```
Comment (MartinSalinas98) [inline tests/unitary/src/compute/test_device_index.cpp:59]: Remove setter

**Review by MartinSalinas98 — CHANGES_REQUESTED** [review]

File: include/xmipp4/core/compute/device_index.inl [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+        // Add to the result
+        result *= base;
+        result += digit;
+
+        ++first;
+    }
+
+    return first;
+}
```
[the remark this answers is not in the document]
Comment (oierlauzi) [reply include/xmipp4/core/compute/device_index.inl]: Done

File: include/xmipp4/core/compute/device_index.inl [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+const std::string& device_index::get_backend_name() const noexcept
+{
+    return m_backend_name;
+}
+
+inline
+void device_index::set_device_id(std::size_t device_index) noexcept
+{
+    m_device_id = device_index;
+}
```
[the remark this answers is not in the document]
Comment (oierlauzi) [reply include/xmipp4/core/compute/device_index.inl]: Done

**Review by MartinSalinas98 — APPROVED** [review]

**Review by MartinSalinas98 — CHANGES_REQUESTED** [review]
Fix Windows warning:
```
error C4996: 'getenv': This function or variable may be unsafe. Consider using _dupenv_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS
```

**Review by MartinSalinas98 — APPROVED** [review]

**Pull request merged**